server/apps/publish/formatters/anpa_formatter.py

```diff
@@ -0,0 +1,119 @@
+            # field seperator
+            anpa += b'\x0A'  # -LF
+            anpa += article.get('priority', 'r').encode('ascii')
+            anpa += b'\x20'
+
+            category = article.get('anpa-category', {})
+            if 'qcode' in category:
+                anpa += article.get('anpa-category', {})['qcode'].encode('ascii')
+            # TODO TODO TODO
+            # NEED TO SORT THIS OUT I THINK THE AUTHORING GETS IT WRONG
+            if 'value' in category:
+                anpa += article.get('anpa-category', {})['value'].encode('ascii')
+
+            anpa += b'\x13'
+            # format identifier
+            if article['type'] == 'preformatted':
+                anpa += b'\x12'
+            else:
+                anpa += b'\x11'
+            anpa += b'\x20'
+
+            # keyword
+            keyword = 'bc-{}'.format(article.get('slugline', '')).replace(' ', '-')
+            keyword = keyword[:24] if len(keyword) > 24 else keyword
+            anpa += keyword.encode('ascii')
+            anpa += b'\x20'
+
+            # version field
+            anpa += b'\x20'
+
+            # reference field
+            anpa += b'\x20'
+
+            # filing date
+            anpa += '{}-{}'.format(article['_updated'].strftime('%m'),
+                                   article['_updated'].strftime('%d')).encode('ascii')
+            anpa += b'\x20'
+
+            # add the word count
+            anpa += str(article.get('word_count', '0000')).zfill(4).encode('ascii')
+            anpa += b'\x0D\x0A'
+
+            anpa += b'\x02'  # STX
+
+            # Add the actual story bits
+            anpa += article.get('headline', '').encode('ascii', 'replace')
+            anpa += b'\x0D\x0A'
+
+            keyword = article.get('slugline', '').encode('ascii', 'ignore')
+            anpa += keyword
+            take_key = article.get('anpa_take_key', '').encode('ascii', 'ignore')
+            anpa += (b'\x20' + take_key) if len(take_key) > 0 else b''
+            anpa += b'\x0D\x0A'
+
+            if article['type'] == 'preformatted':
+                anpa += article.get('body_html', '').encode('ascii', 'replace')
+            else:
+                anpa += BeautifulSoup(article.get('body_html', '')).text.encode('ascii', 'replace')
+
+            anpa += b'\x0D\x0A'
+            if article.get('more_coming', False):
+                anpa += 'MORE'.encode('ascii')
+            else:
+                anpa += article.get('source', '').encode('ascii')
+            sign_off = article.get('sign_off', '').encode('ascii')
+            anpa += (b'\x20' + sign_off) if len(sign_off) > 0 else b''
+            anpa += b'\x0D\x0A'
+
+            anpa += b'\x03'  # ETX
+
+            # time and date
+            anpa += datetime.datetime.now().strftime('%d-%m-%y %H-%M-%S').encode('ascii')
+
+            anpa += b'\x04'  # EOT
+            anpa += b'\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A'
```

Comment:
Repeatedly appending new chunks to an existing string is a slow (`O(n^2)`) operation. It's much better to use `bytearray`, a _mutable_ binary string data type for that ([official docs](https://docs.python.org/3/faq/programming.html#what-is-the-most-efficient-way-to-concatenate-many-strings-together)).